src/backend/optimizer/path/allpaths.c

```diff
@@ -2135,6 +2135,12 @@ set_cte_pathlist(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
 		 */
 		subplan = share_prepared_plan(cteroot, cteplaninfo->shared_plan);
 		subroot = cteplaninfo->subroot;
+
+		if(subplan->flow->locustype == CdbLocusType_General)
+		{
+			subplan->flow->locustype = CdbLocusType_SingleQE;
+			subplan->flow->flotype = FLOW_SINGLETON;
+		}
```

Comment:
We know that Shared Scan producer in case of Hash Join [will allways be](https://github.com/arenadata/gpdb/blob/fcfac650cc8349ece9662dfeea6a34071f8d0db7/src/backend/cdb/cdbmutate.c#L1905-L1910) at inner part of join.
However, what about a case when outer tree is executed at coordinator, and inner tree at segments? At current version of your patch deadlock occurs (generate_series have General locus too):
```
 explain (costs off)
with cte as (
    select c1 from generate_series(1,2) c1
) select * from cte a join (select * from d join cte using(c1) limit 1) b using(c1);

                                              QUERY PLAN                                               
-------------------------------------------------------------------------------------------------------
 Hash Join
   Hash Cond: (share0_ref2.c1 = b.c1)
   ->  Shared Scan (share slice:id 0:0)
   ->  Hash
         ->  Subquery Scan on b
               ->  Limit
                     ->  Gather Motion 3:1  (slice2; segments: 3)
                           ->  Limit
                                 ->  Hash Join
                                       Hash Cond: (d.c1 = share0_ref1.c1)
                                       ->  Seq Scan on d
                                       ->  Hash
                                             ->  Redistribute Motion 1:3  (slice1)
                                                   Hash Key: share0_ref1.c1
                                                   ->  Shared Scan (share slice:id 1:0)
                                                         ->  Materialize
                                                               ->  Function Scan on generate_series c1
 Optimizer: Postgres query optimizer
(18 rows)

```

Reply:
The issue stands not only for General locus but for SingleQE as well:
```
explain (costs off)
with cte as (
    select c1 from d limit 5
) select * from cte a join (select * from d join cte using(c1) limit 1) b using(c1);
                                                    QUERY PLAN                                                    
------------------------------------------------------------------------------------------------------------------
 Hash Join
   Hash Cond: (share0_ref2.c1 = b.c1)
   ->  Shared Scan (share slice:id 0:0)
   ->  Hash
         ->  Subquery Scan on b
               ->  Limit
                     ->  Gather Motion 3:1  (slice3; segments: 3)
                           ->  Limit
                                 ->  Hash Join
                                       Hash Cond: (d.c1 = share0_ref1.c1)
                                       ->  Seq Scan on d
                                       ->  Hash
                                             ->  Redistribute Motion 1:3  (slice2)
                                                   Hash Key: share0_ref1.c1
                                                   ->  Shared Scan (share slice:id 2:0)
                                                         ->  Materialize
                                                               ->  Limit
                                                                     ->  Gather Motion 3:1  (slice1; segments: 3)
                                                                           ->  Limit
                                                                                 ->  Seq Scan on d d_1
 Optimizer: Postgres query optimizer
(21 rows)

```

Reply:
> According to the plans, it seems that `Shared Scan (share slice:id 1:0)` and `Shared Scan (share slice:id 2:0)` should be executed on the entry-db, why is this not happening?

SingleQE doesn't always mean entryDB

Reply:
> According to the plans, it seems that `Shared Scan (share slice:id 1:0)` and `Shared Scan (share slice:id 2:0)` should be executed on the entry-db, why is this not happening?

It seems so. ps output shows slice2 is stuck at coordinator

Reply:
Producer [waits](https://github.com/arenadata/gpdb/blob/fcfac650cc8349ece9662dfeea6a34071f8d0db7/src/backend/executor/nodeShareInputScan.c#L870-L878) until consumers send the ack message

Reply:
On 7X second query can be executed:
```sql
with cte as (
    select c1 from d limit 5
) select * from cte a join (select * from d join cte using(c1) limit 1) b using(c1);
                                                      QUERY PLAN                                                      
----------------------------------------------------------------------------------------------------------------------
 Hash Join (actual time=2.052..2.420 rows=2 loops=1)
   Hash Cond: (d.c1 = share0_ref1.c1)
   Extra Text: Hash chain length 2.0 avg, 2 max, using 1 of 131072 buckets.
   ->  Limit (actual time=0.421..0.427 rows=1 loops=1)
         ->  Gather Motion 3:1  (slice1; segments: 3) (actual time=0.418..0.421 rows=1 loops=1)
               ->  Limit (actual time=1.596..1.855 rows=1 loops=1)
                     ->  Hash Join (actual time=1.592..1.601 rows=1 loops=1)
                           Hash Cond: (d.c1 = share0_ref2.c1)
                           Extra Text: (seg0)   Hash chain length 2.0 avg, 2 max, using 1 of 131072 buckets.
                           ->  Seq Scan on d (actual time=0.019..0.021 rows=1 loops=1)
                           ->  Hash (actual time=1.092..1.096 rows=2 loops=1)
                                 Buckets: 131072  Batches: 1  Memory Usage: 1025kB
                                 ->  Redistribute Motion 1:3  (slice2) (actual time=1.081..1.085 rows=2 loops=1)
                                       Hash Key: share0_ref2.c1
                                       ->  Shared Scan (share slice:id 2:0) (actual time=1.314..1.319 rows=2 loops=1)
   ->  Hash (actual time=1.382..1.390 rows=2 loops=1)
         Buckets: 131072  Batches: 1  Memory Usage: 1025kB
         ->  Shared Scan (share slice:id 0:0) (actual time=1.369..1.379 rows=2 loops=1)
               ->  Limit (actual time=1.289..1.301 rows=2 loops=1)
                     ->  Gather Motion 3:1  (slice3; segments: 3) (actual time=1.286..1.290 rows=2 loops=1)
                           ->  Limit (actual time=0.018..0.029 rows=2 loops=1)
                                 ->  Seq Scan on d d_1 (actual time=0.016..0.020 rows=2 loops=1)
 Optimizer: Postgres-based planner
 Planning Time: 0.758 ms
   (slice0)    Executor memory: 1167K bytes.  Work_mem: 1025K bytes max.
   (slice1)    Executor memory: 1055K bytes avg x 3 workers, 1080K bytes max (seg0).  Work_mem: 1025K bytes max.
   (slice2)    Executor memory: 58K bytes (entry db).
   (slice3)    Executor memory: 39K bytes avg x 3 workers, 39K bytes max (seg0).
 Memory used:  128000kB
 Execution Time: 8.190 ms
(30 rows)
```